lib/TPP/GPU/GpuToCuda.cpp

```diff
@@ -0,0 +1,101 @@
+//===- GpuToCuda.cpp ---------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Passes.h"
+
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/NVVMDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/NVGPU/IR/NVGPUDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/Passes.h"
+
+#include "TPP/Dialect/Tpp/TppDialect.h"
+#include "TPP/PassUtils.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+// Lower generic GPU ops to CUDA backend.
+struct GpuToCuda : public GpuToCudaBase<GpuToCuda>,
+                   UtilityPassBase<gpu::GPUModuleOp> {
+  GpuToCuda() = default;
+  GpuToCuda(StringRef gpuTriple, StringRef gpuChip, StringRef gpuFeatures) {
+    this->gpuTriple = gpuTriple.str();
+    this->gpuChip = gpuChip.str();
+    this->gpuFeatures = gpuFeatures.str();
+  };
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<gpu::GPUDialect>();
+    registry.insert<NVVM::NVVMDialect>();
+    registry.insert<nvgpu::NVGPUDialect>();
+  }
+
+  void runOnOperation() override {
+    auto module = getOperation();
+
+    // Initialize the pipeline if needed.
+    // Otherwise, just run the cached one.
+    if (pm.empty())
+      constructPipeline();
+
+    if (failed(runPipeline(pm, module)))
+      return signalPassFailure();
+  }
+
+private:
+  void constructPipeline() override {
+    pm.clear();
+
+#ifdef TPP_GPU_ENABLE
```

Comment:
I don't like this, but I understand why you did it.

I think a cleaner solution is to not even try to build the library if the GPU support is disabled and then you don't have to `ifdef` anything inside the code.

IIUC, this will clash with `tpp-run` when trying to conditionally add the GPU option, so not a perfect solution, still need figuring that out.

Reply:
In general, most passes are always availables and to support tablegen generated code we need to provide minimum boilerplate. If we don't build anything, we just shift this problem to callers to figure out (more `ifdef`s) when `createGpuToCudaPass` is available.

Macros wouldn't be necessary if all passes followed that logic but `createGpuSerializeToCubinPass` is not available without `MLIR_GPU_TO_CUBIN_PASS_ENABLE` so we need to somehow handle that.
I tried a few alternative approaches but IMO this solution i.e., minimal pass skeleton that does nothing when GPU is disable, is the simplest.

Reply:
Let's go with this for now and decide how to improve later.